<a href="https://colab.research.google.com/github/helloitsdaksh/Pandas_A-Z_cheatsheet/blob/main/Pandas_Cheatsheet.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import pandas as pd

# Creating, Reading and Writing

## DataFrame

In [ ]:
pd.DataFrame({'Yes': [50, 21], 'No': [131, 2]})

Yes   No
0   50  131
1   21    2

In [ ]:
pd.DataFrame({'Bob': ['I liked it.', 'It was awful.'], 
              'Sue': ['Pretty good.', 'Bland.']},
             index=['Product A', 'Product B'])

Bob           Sue
Product A    I liked it.  Pretty good.
Product B  It was awful.        Bland.

## Series 

In [ ]:
pd.Series([0,1,2,3,4,5])

0    0
1    1
2    2
3    3
4    4
5    5
dtype: int64

In [ ]:
pd.Series([30, 35, 40], index=['2015 Sales', '2016 Sales', '2017 Sales'], name='Product A')

2015 Sales    30
2016 Sales    35
2017 Sales    40
Name: Product A, dtype: int64

## Reading

In [ ]:
df = pd.read_csv("/content/sample_data/california_housing_test.csv",) # specify `index_col` if required

In [ ]:
df.shape

(3000, 9)

In [ ]:
df.head()

longitude  latitude  ...  median_income  median_house_value
0    -122.05     37.37  ...         6.6085            344700.0
1    -118.30     34.26  ...         3.5990            176500.0
2    -117.81     33.78  ...         5.7934            270500.0
3    -118.36     33.82  ...         6.1359            330000.0
4    -119.67     36.33  ...         2.9375             81700.0

[5 rows x 9 columns]

In [ ]:
df.describe #df.info

<bound method NDFrame.describe of       longitude  latitude  ...  median_income  median_house_value
0       -122.05     37.37  ...         6.6085            344700.0
1       -118.30     34.26  ...         3.5990            176500.0
2       -117.81     33.78  ...         5.7934            270500.0
3       -118.36     33.82  ...         6.1359            330000.0
4       -119.67     36.33  ...         2.9375             81700.0
...         ...       ...  ...            ...                 ...
2995    -119.86     34.42  ...         1.1790            225000.0
2996    -118.14     34.06  ...         3.3906            237200.0
2997    -119.70     36.30  ...         2.2895             62000.0
2998    -117.12     34.10  ...         3.2708            162500.0
2999    -119.63     34.42  ...         8.5608            500001.0

[3000 rows x 9 columns]>

# Indexing, Selecting & Assigning

##Native accessors 

In [ ]:
df.longitude # or df["longitude"]

0      -122.05
1      -118.30
2      -117.81
3      -118.36
4      -119.67
         ...  
2995   -119.86
2996   -118.14
2997   -119.70
2998   -117.12
2999   -119.63
Name: longitude, Length: 3000, dtype: float64

In [ ]:
df.longitude[0] # or df["lonfitude"][0]

-122.05

## Indexing

Both loc and iloc are row-first, column-second.hence `iloc[r,c]` or `loc[r,c]`
* `iloc` is used for index based selection
* `loc` is used for labeled based selection

### Index based selection

In [ ]:
df.iloc[0]

longitude               -122.0500
latitude                  37.3700
housing_median_age        27.0000
total_rooms             3885.0000
total_bedrooms           661.0000
population              1537.0000
households               606.0000
median_income              6.6085
median_house_value    344700.0000
Name: 0, dtype: float64

In [ ]:
df.iloc[:,0]

0      -122.05
1      -118.30
2      -117.81
3      -118.36
4      -119.67
         ...  
2995   -119.86
2996   -118.14
2997   -119.70
2998   -117.12
2999   -119.63
Name: longitude, Length: 3000, dtype: float64

In [ ]:
df.iloc[:3,0]

0   -122.05
1   -118.30
2   -117.81
Name: longitude, dtype: float64

In [ ]:
# can even pass list
df.iloc[[1,2,3],0]

1   -118.30
2   -117.81
3   -118.36
Name: longitude, dtype: float64

### Label-based selection

In [ ]:
df.loc[0,'longitude']

-122.05

In [ ]:
df.loc[:5,['longitude','latitude']]

longitude  latitude
0    -122.05     37.37
1    -118.30     34.26
2    -117.81     33.78
3    -118.36     33.82
4    -119.67     36.33
5    -119.56     36.51

**NOTE:** 
Choosing between loc and iloc
When choosing or transitioning between loc and iloc, there is one "gotcha" worth keeping in mind, which is that the two methods use slightly different indexing schemes.

iloc uses the Python stdlib indexing scheme, where the first element of the range is included and the last one excluded. So 0:10 will select entries 0,...,9. loc, meanwhile, indexes inclusively. So 0:10 will select entries 0,...,10.

Why the change? Remember that loc can index any stdlib type: strings, for example. If we have a DataFrame with index values Apples, ..., Potatoes, ..., and we want to select "all the alphabetical fruit choices between Apples and Potatoes", then it's a lot more convenient to index df.loc['Apples':'Potatoes'] than it is to index something like df.loc['Apples', 'Potatoet'] (t coming after s in the alphabet).

**This is particularly confusing when the DataFrame index is a simple numerical list, e.g. 0,...,1000. In this case df.iloc[0:1000] will return 1000 entries, while df.loc[0:1000] return 1001 of them! To get 1000 elements using loc, you will need to go one lower and ask for df.loc[0:999].**


Otherwise, the semantics of using loc are the same as those for iloc.

## Manipulating the Index

`set_index` is used to set index as a particular column

In [ ]:
df.set_index("longitude")

latitude  housing_median_age  ...  median_income  median_house_value
longitude                                ...                                   
-122.05       37.37                27.0  ...         6.6085            344700.0
-118.30       34.26                43.0  ...         3.5990            176500.0
-117.81       33.78                27.0  ...         5.7934            270500.0
-118.36       33.82                28.0  ...         6.1359            330000.0
-119.67       36.33                19.0  ...         2.9375             81700.0
...             ...                 ...  ...            ...                 ...
-119.86       34.42                23.0  ...         1.1790            225000.0
-118.14       34.06                27.0  ...         3.3906            237200.0
-119.70       36.30                10.0  ...         2.2895             62000.0
-117.12       34.10                40.0  ...         3.2708            162500.0
-119.63       34.42                42.0  ...         8.5608            500001.0

[3000 rows x 8 columns]

## Conditional Selection

In [ ]:
df.longitude == -122.05

0        True
1       False
2       False
3       False
4       False
        ...  
2995    False
2996    False
2997    False
2998    False
2999    False
Name: longitude, Length: 3000, dtype: bool

In [ ]:
df.loc[df.longitude == -122.05]

longitude  latitude  ...  median_income  median_house_value
0       -122.05     37.37  ...         6.6085            344700.0
73      -122.05     37.36  ...         4.1369            316900.0
83      -122.05     38.26  ...         3.0912            126300.0
1013    -122.05     37.05  ...         4.1679            201300.0
1216    -122.05     37.93  ...         4.4500            184900.0
2370    -122.05     37.31  ...         8.0727            455500.0
2777    -122.05     37.38  ...         4.0720            364200.0

[7 rows x 9 columns]

We can use the ampersand (`&`) to bring the two questions together:

In [ ]:
df.loc[(df.longitude == -122.05) & (df.latitude == 37.37)]

longitude  latitude  ...  median_income  median_house_value
0    -122.05     37.37  ...         6.6085            344700.0

[1 rows x 9 columns]

We can use pip (`|`) to bring or condition i.e either this or that

In [ ]:
df.loc[(df.longitude == -122.05) | (df.latitude == 37.37)]

longitude  latitude  ...  median_income  median_house_value
0       -122.05     37.37  ...         6.6085            344700.0
73      -122.05     37.36  ...         4.1369            316900.0
83      -122.05     38.26  ...         3.0912            126300.0
307     -121.88     37.37  ...         3.5745            333700.0
506     -122.13     37.37  ...        11.8060            500001.0
552     -121.81     37.37  ...         5.1099            213100.0
710     -122.12     37.37  ...        10.7355            500001.0
879     -122.11     37.37  ...         7.2045            500001.0
1013    -122.05     37.05  ...         4.1679            201300.0
1216    -122.05     37.93  ...         4.4500            184900.0
1220    -121.98     37.37  ...         5.0942            217500.0
1578    -121.95     37.37  ...         3.0357            208300.0
2370    -122.05     37.31  ...         8.0727            455500.0
2430    -120.50     37.37  ...         2.4012             79700.0
2566    -121.83     37.37  ...         3.2279            182400.0
2777    -122.05     37.38  ...         4.0720            364200.0

[16 rows x 9 columns]

**NOTE:** You can as many `&` and `|` in a single condition

In [ ]:
df.loc[(df.longitude == -122.05) | ((df.latitude == 37.37) & (df.housing_median_age == 27.0))]

longitude  latitude  ...  median_income  median_house_value
0       -122.05     37.37  ...         6.6085            344700.0
73      -122.05     37.36  ...         4.1369            316900.0
83      -122.05     38.26  ...         3.0912            126300.0
1013    -122.05     37.05  ...         4.1679            201300.0
1216    -122.05     37.93  ...         4.4500            184900.0
2370    -122.05     37.31  ...         8.0727            455500.0
2777    -122.05     37.38  ...         4.0720            364200.0

[7 rows x 9 columns]

In [ ]:
df.loc[(df.longitude == -122.05) & (df.latitude == 37.37) & (df.housing_median_age == 27.0)]

longitude  latitude  ...  median_income  median_house_value
0    -122.05     37.37  ...         6.6085            344700.0

[1 rows x 9 columns]

Pandas comes with a few built-in conditional selectors, two of which we will highlight here.

The first is `isin`. isin is lets you select data whose value "is in" a list of values.

In [ ]:
df.loc[df.longitude.isin([-122.05,-122.13])]

longitude  latitude  ...  median_income  median_house_value
0       -122.05     37.37  ...         6.6085            344700.0
73      -122.05     37.36  ...         4.1369            316900.0
83      -122.05     38.26  ...         3.0912            126300.0
506     -122.13     37.37  ...        11.8060            500001.0
579     -122.13     40.01  ...         2.1250             63300.0
712     -122.13     37.72  ...         7.9120            367300.0
901     -122.13     37.46  ...         3.2642            185600.0
1013    -122.05     37.05  ...         4.1679            201300.0
1216    -122.05     37.93  ...         4.4500            184900.0
1264    -122.13     37.15  ...         3.9423            180300.0
1670    -122.13     37.43  ...         5.2114            438400.0
2063    -122.13     37.70  ...         2.6321            223100.0
2370    -122.05     37.31  ...         8.0727            455500.0
2777    -122.05     37.38  ...         4.0720            364200.0

[14 rows x 9 columns]

The second is `isnull()` (and its companion `notnull()`). These methods let you highlight values which are (or are not) empty (NaN).

In [ ]:
df.loc[df.longitude.notnull()]

longitude  latitude  ...  median_income  median_house_value
0       -122.05     37.37  ...         6.6085            344700.0
1       -118.30     34.26  ...         3.5990            176500.0
2       -117.81     33.78  ...         5.7934            270500.0
3       -118.36     33.82  ...         6.1359            330000.0
4       -119.67     36.33  ...         2.9375             81700.0
...         ...       ...  ...            ...                 ...
2995    -119.86     34.42  ...         1.1790            225000.0
2996    -118.14     34.06  ...         3.3906            237200.0
2997    -119.70     36.30  ...         2.2895             62000.0
2998    -117.12     34.10  ...         3.2708            162500.0
2999    -119.63     34.42  ...         8.5608            500001.0

[3000 rows x 9 columns]

In [ ]:
df.loc[df.longitude.isnull()]

Empty DataFrame
Columns: [longitude, latitude, housing_median_age, total_rooms, total_bedrooms, population, households, median_income, median_house_value]
Index: []

this shows that our database has no NaN in longitude column

## Assigning 

In [ ]:
df['score'] = 'value'

In [ ]:
df.head()

longitude  latitude  ...  median_house_value  score
0    -122.05     37.37  ...            344700.0  value
1    -118.30     34.26  ...            176500.0  value
2    -117.81     33.78  ...            270500.0  value
3    -118.36     33.82  ...            330000.0  value
4    -119.67     36.33  ...             81700.0  value

[5 rows x 10 columns]

In [ ]:
df['index_backwards'] = range(len(df), 0, -1)

In [ ]:
df.head()

longitude  latitude  ...  score  index_backwards
0    -122.05     37.37  ...  value             3000
1    -118.30     34.26  ...  value             2999
2    -117.81     33.78  ...  value             2998
3    -118.36     33.82  ...  value             2997
4    -119.67     36.33  ...  value             2996

[5 rows x 11 columns]

In [ ]:
# df.drop(['score','index_backwards'],axis = 1)

In [ ]:
# axis = 0  is rows
# axis = 1 is columns

#Summary Functions and Maps

##Functions

 Function to get summary of the data are:
 * `describe()`
 * `info()`
 
 other basic function are:
 * `mean()`, `mode()`, `median()`
 * `unique()`
 * `value_counts()`


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   longitude           3000 non-null   float64
 1   latitude            3000 non-null   float64
 2   housing_median_age  3000 non-null   float64
 3   total_rooms         3000 non-null   float64
 4   total_bedrooms      3000 non-null   float64
 5   population          3000 non-null   float64
 6   households          3000 non-null   float64
 7   median_income       3000 non-null   float64
 8   median_house_value  3000 non-null   float64
 9   score               3000 non-null   object 
 10  index_backwards     3000 non-null   int64  
dtypes: float64(9), int64(1), object(1)
memory usage: 257.9+ KB


In [ ]:
# for particular column
df.longitude.describe()

count    3000.000000
mean     -119.589200
std         1.994936
min      -124.180000
25%      -121.810000
50%      -118.485000
75%      -118.020000
max      -114.490000
Name: longitude, dtype: float64

In [ ]:
df.total_rooms.mean()

2599.578666666667

In [ ]:
df.total_rooms.unique()

array([3885., 1510., 3589., ..., 4867.,  573., 5257.])

In [ ]:
df.longitude.value_counts()

-118.21    26
-118.26    26
-118.29    25
-118.28    25
-118.27    25
           ..
-120.33     1
-119.00     1
-119.08     1
-114.61     1
-119.86     1
Name: longitude, Length: 607, dtype: int64

## Map

In this section we will see 
* `map()`
* `apply()`

In [ ]:
longitude_mean = df.longitude.mean()
df.longitude.map(lambda x:x-longitude_mean)

0      -2.4608
1       1.2892
2       1.7792
3       1.2292
4      -0.0808
         ...  
2995   -0.2708
2996    1.4492
2997   -0.1108
2998    2.4692
2999   -0.0408
Name: longitude, Length: 3000, dtype: float64

In [ ]:
def remean_longitude(df):
    df.longitude = df.longitude - longitude_mean
    return df

df.apply(remean_longitude, axis='columns')

longitude  latitude  ...  score  index_backwards
0       -2.4608     37.37  ...  value             3000
1        1.2892     34.26  ...  value             2999
2        1.7792     33.78  ...  value             2998
3        1.2292     33.82  ...  value             2997
4       -0.0808     36.33  ...  value             2996
...         ...       ...  ...    ...              ...
2995    -0.2708     34.42  ...  value                5
2996     1.4492     34.06  ...  value                4
2997    -0.1108     36.30  ...  value                3
2998     2.4692     34.10  ...  value                2
2999    -0.0408     34.42  ...  value                1

[3000 rows x 11 columns]

**NOTE:** If we had called `df.apply()` with `axis='index'`, then instead of passing a function to transform each row, we would need to give a function to transform each column.

In [ ]:
df.total_rooms.astype(str) + " - " + df.latitude.astype(str)

0       3885.0 - 37.37
1       1510.0 - 34.26
2       3589.0 - 33.78
3         67.0 - 33.82
4       1241.0 - 36.33
             ...      
2995    1450.0 - 34.42
2996    5257.0 - 34.06
2997      956.0 - 36.3
2998       96.0 - 34.1
2999    1765.0 - 34.42
Length: 3000, dtype: object

These operators are faster than `map()` or `apply()` because they use speed ups built into pandas. All of the standard Python operators (`>`, `<`, `==`, and so on) work in this manner.

However, they are not as flexible as `map()` or `apply()`, which can do more advanced things, like applying conditional logic, which cannot be done with addition and subtraction alone.

# Grouping and Sorting



## Groupwise analysis

In [ ]:
df = pd.read_csv("/content/sample_data/california_housing_test.csv",) 
df.head()


longitude  latitude  ...  median_income  median_house_value
0    -122.05     37.37  ...         6.6085            344700.0
1    -118.30     34.26  ...         3.5990            176500.0
2    -117.81     33.78  ...         5.7934            270500.0
3    -118.36     33.82  ...         6.1359            330000.0
4    -119.67     36.33  ...         2.9375             81700.0

[5 rows x 9 columns]

In [ ]:
# One function we've been using heavily thus far is the value_counts() function. We can replicate what value_counts() does by doing the following:
df.groupby('longitude').longitude.count()

longitude
-124.18    1
-124.17    1
-124.16    4
-124.15    1
-124.14    3
          ..
-114.98    1
-114.62    1
-114.61    1
-114.55    1
-114.49    1
Name: longitude, Length: 607, dtype: int64

In [ ]:
# print longitude for the total_rooms to be minimum
df.groupby('longitude').total_rooms.min()

longitude
-124.18    1398.0
-124.17    2739.0
-124.16    1668.0
-124.15    2127.0
-124.14    1148.0
            ...  
-114.98    1183.0
-114.62      18.0
-114.61    1187.0
-114.55    2570.0
-114.49    2809.0
Name: total_rooms, Length: 607, dtype: float64

In [ ]:
df.groupby('longitude').apply(lambda x: x.total_rooms)

longitude      
-124.18    1876    1398.0
-124.17    1422    2739.0
-124.16    111     2141.0
           910     1668.0
           1118    2167.0
                    ...  
-114.98    244     1183.0
-114.62    2640      18.0
-114.61    1145    1187.0
-114.55    1595    2570.0
-114.49    2055    2809.0
Name: total_rooms, Length: 3000, dtype: float64

In [ ]:
df.groupby(['longitude', 'latitude']).apply(lambda x: x.loc[x.housing_median_age.idxmax()])

longitude  latitude  ...  median_income  median_house_value
longitude latitude                       ...                                   
-124.18   40.79       -124.18     40.79  ...         1.4668             64600.0
-124.17   41.80       -124.17     41.80  ...         3.7557            109400.0
-124.16   40.77       -124.16     40.77  ...         2.8529             85600.0
          40.79       -124.16     40.79  ...         2.1135             90600.0
          40.80       -124.16     40.80  ...         1.6111             74700.0
...                       ...       ...  ...            ...                 ...
-114.98   33.07       -114.98     33.07  ...         3.1607             57500.0
-114.62   33.62       -114.62     33.62  ...         0.5360            275000.0
-114.61   33.62       -114.61     33.62  ...         2.1759             61500.0
-114.55   32.80       -114.55     32.80  ...         1.2750             56100.0
-114.49   33.97       -114.49     33.97  ...         1.6154             87500.0

[2712 rows x 9 columns]

`agg()` is similar to `apply()`

In [ ]:
df.groupby(['median_house_value']).median_income.agg([len, min, max])

len     min      max
median_house_value                      
22500.0               1  1.0918   1.0918
37500.0               1  1.7969   1.7969
39200.0               1  2.1771   2.1771
39800.0               1  1.9107   1.9107
40000.0               1  3.6875   3.6875
...                 ...     ...      ...
494700.0              1  5.8587   5.8587
495500.0              1  5.9683   5.9683
495800.0              1  4.3125   4.3125
500000.0              4  4.4000   6.7770
500001.0            125  0.4999  15.0001

[1784 rows x 3 columns]

## Multi Indexes

In [ ]:
df2 = df.groupby(['longitude', 'latitude']).housing_median_age.agg([len])
df2

len
longitude latitude     
-124.18   40.79       1
-124.17   41.80       1
-124.16   40.77       1
          40.79       1
          40.80       1
...                 ...
-114.98   33.07       1
-114.62   33.62       1
-114.61   33.62       1
-114.55   32.80       1
-114.49   33.97       1

[2712 rows x 1 columns]

In [ ]:
df2.index

MultiIndex([(-124.18, 40.79),
            (-124.17,  41.8),
            (-124.16, 40.77),
            (-124.16, 40.79),
            (-124.16,  40.8),
            (-124.16, 41.92),
            (-124.15, 40.78),
            (-124.14,  40.6),
            (-124.14, 40.72),
            (-124.14,  40.8),
            ...
            (-115.57, 32.79),
            (-115.57,  32.8),
            (-115.56, 32.78),
            (-115.52, 32.98),
            (-115.49, 32.67),
            (-114.98, 33.07),
            (-114.62, 33.62),
            (-114.61, 33.62),
            (-114.55,  32.8),
            (-114.49, 33.97)],
           names=['longitude', 'latitude'], length=2712)

In [ ]:
df3 = df2.reset_index()
df3

longitude  latitude  len
0       -124.18     40.79    1
1       -124.17     41.80    1
2       -124.16     40.77    1
3       -124.16     40.79    1
4       -124.16     40.80    1
...         ...       ...  ...
2707    -114.98     33.07    1
2708    -114.62     33.62    1
2709    -114.61     33.62    1
2710    -114.55     32.80    1
2711    -114.49     33.97    1

[2712 rows x 3 columns]

In [ ]:
df3.head(10)

longitude  latitude  len
0    -124.18     40.79    1
1    -124.17     41.80    1
2    -124.16     40.77    1
3    -124.16     40.79    1
4    -124.16     40.80    1
5    -124.16     41.92    1
6    -124.15     40.78    1
7    -124.14     40.60    1
8    -124.14     40.72    1
9    -124.14     40.80    1

As you can se there was a grouping in the data as there were 3000 records but now its reduced to 2712
However, in general the multi-index method you will use most often is the one for converting back to a regular index, the `reset_index()` method:



## Sorting

using `sort_values()` function

In [ ]:
df3.sort_values(by='len')

longitude  latitude  len
0       -124.18     40.79    1
1759    -118.23     33.92    1
1760    -118.23     33.93    1
1761    -118.23     33.96    1
1762    -118.23     34.09    1
...         ...       ...  ...
192     -122.42     37.76    3
177     -122.43     37.73    3
1694    -118.27     33.94    4
1664    -118.29     34.05    4
325     -122.25     37.80    4

[2712 rows x 3 columns]

In [ ]:
df3.sort_values(by='len',ascending = False)

longitude  latitude  len
325     -122.25     37.80    4
1664    -118.29     34.05    4
1694    -118.27     33.94    4
192     -122.42     37.76    3
2444    -117.25     32.80    3
...         ...       ...  ...
950     -121.00     37.64    1
951     -121.00     37.66    1
952     -120.99     37.61    1
953     -120.99     37.70    1
2711    -114.49     33.97    1

[2712 rows x 3 columns]

using `sort_index()` function 

In [ ]:
df3.sort_index()

longitude  latitude  len
0       -124.18     40.79    1
1       -124.17     41.80    1
2       -124.16     40.77    1
3       -124.16     40.79    1
4       -124.16     40.80    1
...         ...       ...  ...
2707    -114.98     33.07    1
2708    -114.62     33.62    1
2709    -114.61     33.62    1
2710    -114.55     32.80    1
2711    -114.49     33.97    1

[2712 rows x 3 columns]

In [ ]:
df3.sort_values(by = ['longitude','latitude'])

longitude  latitude  len
0       -124.18     40.79    1
1       -124.17     41.80    1
2       -124.16     40.77    1
3       -124.16     40.79    1
4       -124.16     40.80    1
...         ...       ...  ...
2707    -114.98     33.07    1
2708    -114.62     33.62    1
2709    -114.61     33.62    1
2710    -114.55     32.80    1
2711    -114.49     33.97    1

[2712 rows x 3 columns]

# Data Types and Missing Values

In [ ]:
## Dtypes

In [ ]:
df.longitude.dtype

dtype('float64')

In [ ]:
df.dtypes

longitude             float64
latitude              float64
housing_median_age    float64
total_rooms           float64
total_bedrooms        float64
population            float64
households            float64
median_income         float64
median_house_value    float64
dtype: object

In [ ]:
df.total_rooms.astype('int64')


0       3885
1       1510
2       3589
3         67
4       1241
        ... 
2995    1450
2996    5257
2997     956
2998      96
2999    1765
Name: total_rooms, Length: 3000, dtype: int64

## Missing data

In [ ]:
import numpy as np

In [ ]:
data = {'set_of_numbers': [1,2,3,4,5,np.nan,6,7,np.nan,np.nan,8,9,10,np.nan]}
df = pd.DataFrame(data)

print (df)

    set_of_numbers
0              1.0
1              2.0
2              3.0
3              4.0
4              5.0
5              NaN
6              6.0
7              7.0
8              NaN
9              NaN
10             8.0
11             9.0
12            10.0
13             NaN


In [ ]:
df[pd.isnull(df.set_of_numbers)]

set_of_numbers
5              NaN
8              NaN
9              NaN
13             NaN

In [ ]:
df.fillna('unknown')

set_of_numbers
0             1.0
1             2.0
2             3.0
3             4.0
4             5.0
5         unknown
6             6.0
7             7.0
8         unknown
9         unknown
10            8.0
11            9.0
12           10.0
13        unknown

In [ ]:
df.set_of_numbers.replace('unknown','nan')

0      1.0
1      2.0
2      3.0
3      4.0
4      5.0
5      NaN
6      6.0
7      7.0
8      NaN
9      NaN
10     8.0
11     9.0
12    10.0
13     NaN
Name: set_of_numbers, dtype: float64

In [ ]:
df.fillna(df.set_of_numbers.median())

set_of_numbers
0              1.0
1              2.0
2              3.0
3              4.0
4              5.0
5              5.5
6              6.0
7              7.0
8              5.5
9              5.5
10             8.0
11             9.0
12            10.0
13             5.5

# Renaming and Combining

## Renaming


In [ ]:
df.rename(columns={'total_rooms':'TOTAL_ROOMS'} )

longitude  latitude  ...  median_income  median_house_value
0       -122.05     37.37  ...         6.6085            344700.0
1       -118.30     34.26  ...         3.5990            176500.0
2       -117.81     33.78  ...         5.7934            270500.0
3       -118.36     33.82  ...         6.1359            330000.0
4       -119.67     36.33  ...         2.9375             81700.0
...         ...       ...  ...            ...                 ...
2995    -119.86     34.42  ...         1.1790            225000.0
2996    -118.14     34.06  ...         3.3906            237200.0
2997    -119.70     36.30  ...         2.2895             62000.0
2998    -117.12     34.10  ...         3.2708            162500.0
2999    -119.63     34.42  ...         8.5608            500001.0

[3000 rows x 9 columns]

In [ ]:
df.rename(index={0: 'firstEntry', 1: 'secondEntry'})

longitude  latitude  ...  median_income  median_house_value
firstEntry     -122.05     37.37  ...         6.6085            344700.0
secondEntry    -118.30     34.26  ...         3.5990            176500.0
2              -117.81     33.78  ...         5.7934            270500.0
3              -118.36     33.82  ...         6.1359            330000.0
4              -119.67     36.33  ...         2.9375             81700.0
...                ...       ...  ...            ...                 ...
2995           -119.86     34.42  ...         1.1790            225000.0
2996           -118.14     34.06  ...         3.3906            237200.0
2997           -119.70     36.30  ...         2.2895             62000.0
2998           -117.12     34.10  ...         3.2708            162500.0
2999           -119.63     34.42  ...         8.5608            500001.0

[3000 rows x 9 columns]

You'll probably rename columns very often, but rename index values very rarely. For that, `set_index()` is usually more convenient.

Both the row index and the column index can have their own name attribute. The complimentary `rename_axis()` method may be used to change these names. For example:

In [ ]:
df.rename_axis("houses", axis='rows').rename_axis("fields", axis='columns')

fields  longitude  latitude  ...  median_income  median_house_value
houses                       ...                                   
0         -122.05     37.37  ...         6.6085            344700.0
1         -118.30     34.26  ...         3.5990            176500.0
2         -117.81     33.78  ...         5.7934            270500.0
3         -118.36     33.82  ...         6.1359            330000.0
4         -119.67     36.33  ...         2.9375             81700.0
...           ...       ...  ...            ...                 ...
2995      -119.86     34.42  ...         1.1790            225000.0
2996      -118.14     34.06  ...         3.3906            237200.0
2997      -119.70     36.30  ...         2.2895             62000.0
2998      -117.12     34.10  ...         3.2708            162500.0
2999      -119.63     34.42  ...         8.5608            500001.0

[3000 rows x 9 columns]

## Combining

In [4]:
train = pd.read_csv("/content/sample_data/california_housing_train.csv")
test = pd.read_csv("/content/sample_data/california_housing_test.csv")

pd.concat([train, test])

longitude  latitude  ...  median_income  median_house_value
0       -114.31     34.19  ...         1.4936             66900.0
1       -114.47     34.40  ...         1.8200             80100.0
2       -114.56     33.69  ...         1.6509             85700.0
3       -114.57     33.64  ...         3.1917             73400.0
4       -114.57     33.57  ...         1.9250             65500.0
...         ...       ...  ...            ...                 ...
2995    -119.86     34.42  ...         1.1790            225000.0
2996    -118.14     34.06  ...         3.3906            237200.0
2997    -119.70     36.30  ...         2.2895             62000.0
2998    -117.12     34.10  ...         3.2708            162500.0
2999    -119.63     34.42  ...         8.5608            500001.0

[20000 rows x 9 columns]

In [5]:
test.shape, train.shape

((3000, 9), (17000, 9))

The middlemost combiner in terms of complexity is `join()`. `join()` lets you combine different DataFrame objects which have an index in common.

In [7]:
left = train
right = test

left.join(right, lsuffix='_train', rsuffix='_test')

longitude_train  ...  median_house_value_test
0              -114.31  ...                 344700.0
1              -114.47  ...                 176500.0
2              -114.56  ...                 270500.0
3              -114.57  ...                 330000.0
4              -114.57  ...                  81700.0
...                ...  ...                      ...
16995          -124.26  ...                      NaN
16996          -124.27  ...                      NaN
16997          -124.30  ...                      NaN
16998          -124.30  ...                      NaN
16999          -124.35  ...                      NaN

[17000 rows x 18 columns]